pbft/pbft.cpp

```diff
@@ -1429,14 +1485,57 @@ pbft::handle_viewchange(const pbft_msg &msg, const bzn_envelope &original_msg)
 void
 pbft::handle_newview(const pbft_msg& msg, const bzn_envelope& original_msg)
 {
-    if (!this->is_valid_newview_message(msg, original_msg))
+    // are we just now joining the swarm?
+    if (this->in_swarm == swarm_status::waiting)
+    {
+        if (!validate_config_in_newview(msg))
+        {
+            LOG(debug) << "newview received with unsubstantiated configuration";
+            return;
+        }
+
+        auto newconfig = std::make_shared<pbft_configuration>();
+        if (!newconfig->from_string(msg.config()))
+        {
+            LOG(debug) << "newview received with invalid configuration";
+            return;
+        }
+
+        auto hash = newconfig->get_hash();
+        this->configurations.add(newconfig);
+        this->configurations.enable(hash);
+
+        // we need to switch to this configuration now so we have the peer info to validate the message
+        this->move_to_new_configuration(hash);
+
+        if (!this->is_valid_newview_message(msg, original_msg))
+        {
+            LOG (debug) << "handle_newview - ignoring invalid NEWVIEW message while waiting to join swarm";
+            return;
+        }
+
+        // adopt checkpoint in newview message
+        pbft_msg viewchange;
+        viewchange.ParseFromString(msg.viewchange_messages(0).pbft());
+        this->save_checkpoint(viewchange);
+        this->in_swarm = swarm_status::joined;
```

Comment:
Do we want to change our state if we still might reject the view change?

Reply:
In theory this is benign - the node is not yet a member of the swarm and will not be until it receives a valid NEW_VIEW message. If you think it's worth it I can add code to reset to a null configuration in the case where the new_view is invalid.  Either way, there is a tautology here - the config contains the credentials to validate the new_view but if we don't trust the new_view we can't strictly use the config. I'll put in a comment to that effect and assume we will be retrieving config out-of-band in the future.

Reply:
yeah, i guess now that you point it out, accepting the config is the scary part, and we do need a trust root for that anyway.